src/node.rs

```diff
@@ -729,6 +729,11 @@ impl Node {
             .collect()
     }
 
+    fn should_be_connected(&self, node: Name, blocks: &Blocks) -> bool {
+        let neighbours = nodes_in_any(blocks, &self.current_blocks);
+        neighbours.contains(&node)
```

Comment:
It's a bit dodgy, but we connect to joining nodes no matter what when we get their `NodeJoined` message... There's a `TODO` in the code to change it so that only the relevant nodes get that initial `NodeJoined` message, but I think we can leave that for another PR.